port/win/port_win.h

```diff
@@ -0,0 +1,576 @@
+#  define WIN32_LEAN_AND_MEAN
+#endif
+
+// Assume that for everywhere
+#undef PLATFORM_IS_LITTLE_ENDIAN
+#define PLATFORM_IS_LITTLE_ENDIAN true
+
+#include <windows.h>
+#include <string>
+#include <string.h>
+#include <mutex>
+#include <condition_variable>
+
+#include <stdint.h>
+
+#include "rocksdb/options.h"
+
+#ifndef strcasecmp
+#define strcasecmp _stricmp
+#endif
+
+// defined in stdio.h
+#ifndef snprintf
+#define snprintf _snprintf
+#endif
+
+typedef SSIZE_T ssize_t;
+
+// size_t printf formatting named in the manner of C99 standard formatting strings such as PRIu64
+// in fact, we could use that one
+#define ROCKSDB_PRIszt "Iu"
+
+#define __attribute__(A)
+
+#ifdef ZLIB
+#include <zlib.h>
+#endif
+
+#ifdef BZIP2
+#include <bzlib.h>
+#endif
+
+#if defined(LZ4)
+#include <lz4.h>
+#include <lz4hc.h>
+#endif
+
+#ifdef SNAPPY
+#include <snappy.h>
+#endif
+
+// Thread local storage on Linux
+// There is thread_local in C++11
+#define __thread __declspec(thread)
+
+#ifndef PLATFORM_IS_LITTLE_ENDIAN
+#define PLATFORM_IS_LITTLE_ENDIAN (__BYTE_ORDER == __LITTLE_ENDIAN)
+#endif
+
+namespace rocksdb {
+
+#define PREFETCH(addr, rw, locality)
+
+namespace port 
+{
+
+// For use at db/file_indexer.h kLevelMaxIndex
+const int LevelMaxIndex = INT32_MAX;
+
+const bool kLittleEndian = true;
+
+class CondVar;
+
+class Mutex 
+{
+public:
+    /* implicit */ Mutex(bool adaptive = false);
+    ~Mutex();
+
+    void Lock();
+    void Unlock();
+  
+    // this will assert if the mutex is not locked
+    // it does NOT verify that mutex is held by a calling thread
+    void AssertHeld();
+
+    std::unique_lock<std::mutex>& getLock() {
+        return lock;
+    }
+
+private:
+    friend class CondVar;
+    std::mutex m_mutex;
+    std::unique_lock<std::mutex> lock;
+#ifndef NDEBUG
+    bool locked_;
+#endif
+
+    // No copying
+    Mutex(const Mutex&);
+    void operator=(const Mutex&);
+};
+
+class RWMutex 
+{
+public:
+    RWMutex() {
+      InitializeSRWLock(&srwLock_);
+    }
+
+    void ReadLock() {
+      AcquireSRWLockShared(&srwLock_);
+    }
+
+    void WriteLock() {
+      AcquireSRWLockExclusive(&srwLock_);
+    }
+
+    void ReadUnlock() {
+      ReleaseSRWLockShared(&srwLock_);
+    }
+
+    void WriteUnlock() {
+      ReleaseSRWLockExclusive(&srwLock_);
+    }
+
+    // Empty as in POSIX
+    void AssertHeld() { }
+
+private:
+
+  SRWLOCK srwLock_;
+  // No copying allowed
+  RWMutex(const RWMutex&);
+  void operator=(const RWMutex&);
+};
+
+class CondVar 
+{
+public:
+    explicit CondVar(Mutex* mu);
+    ~CondVar();
+    void Wait();
+    bool TimedWait(uint64_t expiration_time);
+    void Signal();
+    void SignalAll();
+private:
+    std::condition_variable cv_;
+    Mutex * mu_;
+};
+
+typedef std::once_flag OnceType;
+#define LEVELDB_ONCE_INIT std::once_flag::once_flag();
+extern void InitOnce(OnceType* once, void (*initializer)());
+
+inline bool Snappy_Compress(const CompressionOptions& opts, const char* input,
```

Comment:
I'm curious why do you need a separate implementation of compression functions? I thought they're platform independent.